File: cpp/c-api/src/core/local_id_builder.cpp
#include "dnv/vista/sdk/c/core/local_id_builder.h"

#include "local_id_internal.h"

#include "../cast_internal.h"
#include "../error_internal.h"

using namespace dnv::vista::sdk;
using dnv::vista::sdk::c::fromGmodPathValue;
using dnv::vista::sdk::c::fromLocalId;
using dnv::vista::sdk::c::fromLocalIdBuilder;
using dnv::vista::sdk::c::fromLocalIdValue;
using dnv::vista::sdk::c::selectMetadataTag;
using dnv::vista::sdk::c::toCodebookName;
using dnv::vista::sdk::c::toGmodPath;
using dnv::vista::sdk::c::toLocalIdBuilder;
using dnv::vista::sdk::c::toOwnedCString;
using dnv::vista::sdk::c::toTag;

const char* dnv_vista_sdk_local_id_builder_naming_rule(void)
{
    return LocalIdBuilder::namingRule().data();
}

dnv_vista_sdk_local_id_builder_t* dnv_vista_sdk_local_id_builder_create(const char* visVersion)
{
    if (visVersion == nullptr)
    {
        c::setLastErrorMessage("visVersion must not be null");
        return nullptr;
    }

    const auto version = VisVersions::fromString(visVersion);
    if (!version.has_value())
    {
        c::setLastErrorMessage("unrecognized VIS version");
        return nullptr;
    }

    return fromLocalIdBuilder(LocalIdBuilder::create(*version));
}

void dnv_vista_sdk_local_id_builder_free(dnv_vista_sdk_local_id_builder_t* builder)
{
    delete reinterpret_cast<LocalIdBuilder*>(builder);
}

const char* dnv_vista_sdk_local_id_builder_version(const dnv_vista_sdk_local_id_builder_t* builder)
{
    if (builder == nullptr)
    {
        c::setLastErrorMessage("builder must not be null");
        return nullptr;
    }

    const auto version = toLocalIdBuilder(builder)->version();
    if (!version.has_value())
    {
        return nullptr;
    }

    return VisVersions::toString(*version).data();
}

int dnv_vista_sdk_local_id_builder_is_verbose_mode(const dnv_vista_sdk_local_id_builder_t* builder)
{
    if (builder == nullptr)
    {
        c::setLastErrorMessage("builder must not be null");
        return 0;
    }

    return toLocalIdBuilder(builder)->isVerboseMode() ? 1 : 0;
}

int dnv_vista_sdk_local_id_builder_is_valid(const dnv_vista_sdk_local_id_builder_t* builder)
{
    if (builder == nullptr)
    {
        c::setLastErrorMessage("builder must not be null");
        return 0;
    }

    return toLocalIdBuilder(builder)->isValid() ? 1 : 0;
}

int dnv_vista_sdk_local_id_builder_is_empty(const dnv_vista_sdk_local_id_builder_t* builder)
{
    if (builder == nullptr)
    {
        c::setLastErrorMessage("builder must not be null");
        return 0;
    }

    return toLocalIdBuilder(builder)->isEmpty() ? 1 : 0;
}

int dnv_vista_sdk_local_id_builder_is_empty_metadata(const dnv_vista_sdk_local_id_builder_t* builder)
{
    if (builder == nullptr)
    {
        c::setLastErrorMessage("builder must not be null");
        return 0;
    }

    return toLocalIdBuilder(builder)->isEmptyMetadata() ? 1 : 0;
}

int dnv_vista_sdk_local_id_builder_has_custom_tag(const dnv_vista_sdk_local_id_builder_t* builder)
{
    if (builder == nullptr)
    {
        c::setLastErrorMessage("builder must not be null");
        return 0;
    }

    return toLocalIdBuilder(builder)->hasCustomTag() ? 1 : 0;
}

dnv_vista_sdk_gmod_path_t* dnv_vista_sdk_local_id_builder_primary_item(const dnv_vista_sdk_local_id_builder_t* builder)
{
    if (builder == nullptr)
    {
        c::setLastErrorMessage("builder must not be null");
        return nullptr;
    }

    const auto& path = toLocalIdBuilder(builder)->primaryItem();
    if (!path.has_value())
    {
        return nullptr;
    }

    return fromGmodPathValue(GmodPath{ *path });
}

dnv_vista_sdk_gmod_path_t* dnv_vista_sdk_local_id_builder_secondary_item(
    const dnv_vista_sdk_local_id_builder_t* builder)
{
    if (builder == nullptr)
    {
        c::setLastErrorMessage("builder must not be null");
        return nullptr;
    }

    const auto& path = toLocalIdBuilder(builder)->secondaryItem();
    if (!path.has_value())
    {
        return nullptr;
    }

    return fromGmodPathValue(GmodPath{ *path });
}

dnv_vista_sdk_metadata_tag_t* dnv_vista_sdk_local_id_builder_metadata_tag(
    const dnv_vista_sdk_local_id_builder_t* builder, dnv_vista_sdk_codebook_name_t name)
{
    if (builder == nullptr)
    {
        c::setLastErrorMessage("builder must not be null");
        return nullptr;
    }

    const auto* tag = selectMetadataTag(*toLocalIdBuilder(builder), toCodebookName(name));
    if (tag == nullptr)
    {
        c::setLastErrorMessage("invalid codebook name");
        return nullptr;
    }

    if (!tag->has_value())
    {
        return nullptr;
    }

    return reinterpret_cast<dnv_vista_sdk_metadata_tag_t*>(new MetadataTag{ **tag });
}

dnv_vista_sdk_local_id_builder_t* dnv_vista_sdk_local_id_builder_with_vis_version(
    const dnv_vista_sdk_local_id_builder_t* builder, const char* visVersion)
{
    if (builder == nullptr || visVersion == nullptr)
    {
        c::setLastErrorMessage("builder and visVersion must not be null");
        return nullptr;
    }

    try
    {
        return fromLocalIdBuilder(toLocalIdBuilder(builder)->withVisVersion(std::string_view{ visVersion }));
    }
    catch (const std::exception& e)
    {
        c::setLastErrorMessage(e.what());
        return nullptr;
    }
}

dnv_vista_sdk_local_id_builder_t* dnv_vista_sdk_local_id_builder_without_vis_version(
    const dnv_vista_sdk_local_id_builder_t* builder)
{
    if (builder == nullptr)
    {
        c::setLastErrorMessage("builder must not be null");
        return nullptr;
    }

    return fromLocalIdBuilder(toLocalIdBuilder(builder)->withoutVisVersion());
}

dnv_vista_sdk_local_id_builder_t* dnv_vista_sdk_local_id_builder_with_primary_item(
    const dnv_vista_sdk_local_id_builder_t* builder, const dnv_vista_sdk_gmod_path_t* path)
{
    if (builder == nullptr || path == nullptr)
    {
        c::setLastErrorMessage("builder and path must not be null");
        return nullptr;
    }

    return fromLocalIdBuilder(toLocalIdBuilder(builder)->withPrimaryItem(*toGmodPath(path)));
}

dnv_vista_sdk_local_id_builder_t* dnv_vista_sdk_local_id_builder_without_primary_item(
    const dnv_vista_sdk_local_id_builder_t* builder)
{
    if (builder == nullptr)
    {
        c::setLastErrorMessage("builder must not be null");
        return nullptr;
    }

    return fromLocalIdBuilder(toLocalIdBuilder(builder)->withoutPrimaryItem());
}

dnv_vista_sdk_local_id_builder_t* dnv_vista_sdk_local_id_builder_with_secondary_item(
    const dnv_vista_sdk_local_id_builder_t* builder, const dnv_vista_sdk_gmod_path_t* path)
{
    if (builder == nullptr || path == nullptr)
    {
        c::setLastErrorMessage("builder and path must not be null");
        return nullptr;
    }

    return fromLocalIdBuilder(toLocalIdBuilder(builder)->withSecondaryItem(*toGmodPath(path)));
}

dnv_vista_sdk_local_id_builder_t* dnv_vista_sdk_local_id_builder_without_secondary_item(
    const dnv_vista_sdk_local_id_builder_t* builder)
{
    if (builder == nullptr)
    {
        c::setLastErrorMessage("builder must not be null");
        return nullptr;
    }

    return fromLocalIdBuilder(toLocalIdBuilder(builder)->withoutSecondaryItem());
}

dnv_vista_sdk_local_id_builder_t* dnv_vista_sdk_local_id_builder_with_metadata_tag(
    const dnv_vista_sdk_local_id_builder_t* builder, const dnv_vista_sdk_metadata_tag_t* tag)
{
    if (builder == nullptr || tag == nullptr)
    {
        c::setLastErrorMessage("builder and tag must not be null");
        return nullptr;
    }

    try
    {
        return fromLocalIdBuilder(toLocalIdBuilder(builder)->withMetadataTag(*toTag(tag)));
    }
    catch (const std::exception& e)
    {
        c::setLastErrorMessage(e.what());
        return nullptr;
    }
}

dnv_vista_sdk_local_id_builder_t* dnv_vista_sdk_local_id_builder_without_metadata_tag(
    const dnv_vista_sdk_local_id_builder_t* builder, dnv_vista_sdk_codebook_name_t name)
{
    if (builder == nullptr)
    {
        c::setLastErrorMessage("builder must not be null");
        return nullptr;
    }

    return fromLocalIdBuilder(toLocalIdBuilder(builder)->withoutMetadataTag(toCodebookName(name)));
}

dnv_vista_sdk_local_id_builder_t* dnv_vista_sdk_local_id_builder_with_verbose_mode(
    const dnv_vista_sdk_local_id_builder_t* builder, int verbose)
{
    if (builder == nullptr)
    {
        c::setLastErrorMessage("builder must not be null");
        return nullptr;
    }

    return fromLocalIdBuilder(toLocalIdBuilder(builder)->withVerboseMode(verbose != 0));
}

dnv_vista_sdk_local_id_t* dnv_vista_sdk_local_id_builder_build(const dnv_vista_sdk_local_id_builder_t* builder)
{
    if (builder == nullptr)
    {
        c::setLastErrorMessage("builder must not be null");
        return nullptr;
    }

    try
    {
        return fromLocalIdValue(toLocalIdBuilder(builder)->build());
    }
    catch (const std::exception& e)
    {
        c::setLastErrorMessage(e.what());
        return nullptr;
    }
}

char* dnv_vista_sdk_local_id_builder_to_string(const dnv_vista_sdk_local_id_builder_t* builder)
{
    if (builder == nullptr)
    {
        c::setLastErrorMessage("builder must not be null");
        return nullptr;
    }

    return toOwnedCString(toLocalIdBuilder(builder)->toString());
}

dnv_vista_sdk_local_id_t* dnv_vista_sdk_local_id_builder_from_string(const char* localIdStr)
{
    if (localIdStr == nullptr)
    {
        c::setLastErrorMessage("localIdStr must not be null");
        return nullptr;
    }

    auto localId = LocalIdBuilder::fromString(localIdStr);
    if (!localId.has_value())
    {
        c::setLastErrorMessage("invalid LocalId string");
    }

    return fromLocalId(std::move(localId));
}

dnv_vista_sdk_local_id_t* dnv_vista_sdk_local_id_builder_from_string_with_errors(
    const char* localIdStr, dnv_vista_sdk_parsing_errors_t** outErrors)
{
    if (localIdStr == nullptr || outErrors == nullptr)
    {
        c::setLastErrorMessage("localIdStr and outErrors must not be null");
        return nullptr;
    }

    ParsingErrors errors;
    auto localId = LocalIdBuilder::fromString(localIdStr, errors);

    *outErrors = reinterpret_cast<dnv_vista_sdk_parsing_errors_t*>(new ParsingErrors{ std::move(errors) });

    if (!localId.has_value())
    {
        c::setLastErrorMessage("invalid LocalId string");
    }

    return fromLocalId(std::move(localId));
}

void dnv_vista_sdk_local_id_builder_string_free(char* str)
{
    delete[] str;
}

File: bindings/rust/vista-sdk/src/core/local_id_builder.rs
use std::ffi::{CStr, CString};
use std::os::raw::c_int;
use std::ptr;
use std::str::FromStr;

use crate::core::codebook_name::{codebook_names, CodebookName};
use crate::core::error::{last_error, VistaError};
use crate::core::gmod_path::GmodPath;
use crate::core::local_id::LocalId;
use crate::core::metadata_tag::MetadataTag;
use crate::core::parsing_errors::ParsingErrors;
use crate::core::vis_version::VisVersion;
use crate::ffi::core::local_id_builder as ffi;

/// Builder for constructing [`LocalIdRef`](crate::core::local_id::LocalIdRef) instances with a fluent API.
///
/// Borrowed - obtained via accessors like [`LocalIdRef::builder`](crate::core::local_id::LocalIdRef::builder). See
/// [`LocalIdBuilder`] for the owned variant returned by [`LocalIdBuilder::create`].
#[repr(transparent)]
pub struct LocalIdBuilderRef(ffi::dnv_vista_sdk_local_id_builder_t);

impl LocalIdBuilderRef {
    /// # Safety
    /// `ptr` must be non-null and valid for `'a`.
    pub(crate) unsafe fn from_ptr<'a>(
        ptr: *const ffi::dnv_vista_sdk_local_id_builder_t,
    ) -> &'a Self {
        // SAFETY: caller guarantees ptr is non-null and valid. `LocalIdBuilderRef` is `#[repr(transparent)]`.
        unsafe { &*(ptr as *const LocalIdBuilderRef) }
    }

    pub(crate) fn as_ffi_ptr(&self) -> *const ffi::dnv_vista_sdk_local_id_builder_t {
        self as *const Self as *const ffi::dnv_vista_sdk_local_id_builder_t
    }

    /// ISO 19848 Annex C naming rule identifier (`"dnv-v2"`).
    pub fn naming_rule() -> &'static str {
        // SAFETY: no arguments. Returns a non-null, statically-owned pointer.
        let ptr = unsafe { ffi::dnv_vista_sdk_local_id_builder_naming_rule() };
        assert!(
            !ptr.is_null(),
            "dnv_vista_sdk_local_id_builder_naming_rule returned NULL"
        );
        // SAFETY: ptr is non-null and owned by the library for the program's lifetime.
        unsafe { CStr::from_ptr(ptr) }
            .to_str()
            .expect("invalid UTF-8 in local id builder naming rule")
    }

    /// VIS version set on this builder, if any.
    pub fn version(&self) -> Option<VisVersion> {
        // SAFETY: self is non-null and valid for the lifetime of `&self`.
        let ptr = unsafe { ffi::dnv_vista_sdk_local_id_builder_version(self.as_ffi_ptr()) };
        if ptr.is_null() {
            return None;
        }
        // SAFETY: ptr is non-null (checked above) and owned by `self` for its lifetime.
        let s = unsafe { CStr::from_ptr(ptr) }
            .to_str()
            .expect("invalid UTF-8 in local id builder version");
        Some(VisVersion::from_str(s).expect("unrecognized local id builder version string"))
    }

    /// Whether verbose mode is enabled.
    pub fn is_verbose_mode(&self) -> bool {
        // SAFETY: self is non-null and valid for the lifetime of `&self`.
        unsafe { ffi::dnv_vista_sdk_local_id_builder_is_verbose_mode(self.as_ffi_ptr()) != 0 }
    }

    /// Whether the builder state is complete enough to call [`build`](LocalIdBuilderRef::build).
    pub fn is_valid(&self) -> bool {
        // SAFETY: self is non-null and valid for the lifetime of `&self`.
        unsafe { ffi::dnv_vista_sdk_local_id_builder_is_valid(self.as_ffi_ptr()) != 0 }
    }

    /// Whether no components have been set yet.
    pub fn is_empty(&self) -> bool {
        // SAFETY: self is non-null and valid for the lifetime of `&self`.
        unsafe { ffi::dnv_vista_sdk_local_id_builder_is_empty(self.as_ffi_ptr()) != 0 }
    }

    /// Whether no metadata tags have been set.
    pub fn is_empty_metadata(&self) -> bool {
        // SAFETY: self is non-null and valid for the lifetime of `&self`.
        unsafe { ffi::dnv_vista_sdk_local_id_builder_is_empty_metadata(self.as_ffi_ptr()) != 0 }
    }

    /// Whether any metadata tag uses a non-standard value.
    pub fn has_custom_tag(&self) -> bool {
        // SAFETY: self is non-null and valid for the lifetime of `&self`.
        unsafe { ffi::dnv_vista_sdk_local_id_builder_has_custom_tag(self.as_ffi_ptr()) != 0 }
    }

    /// Primary item (Gmod path), if set.
    pub fn primary_item(&self) -> Option<GmodPath> {
        // SAFETY: self is non-null and valid for the lifetime of `&self`.
        let ptr = unsafe { ffi::dnv_vista_sdk_local_id_builder_primary_item(self.as_ffi_ptr()) };
        if ptr.is_null() {
            None
        } else {
            Some(GmodPath::from_owned_ptr(ptr))
        }
    }

    /// Secondary item (Gmod path), if set.
    pub fn secondary_item(&self) -> Option<GmodPath> {
        // SAFETY: self is non-null and valid for the lifetime of `&self`.
        let ptr = unsafe { ffi::dnv_vista_sdk_local_id_builder_secondary_item(self.as_ffi_ptr()) };
        if ptr.is_null() {
            None
        } else {
            Some(GmodPath::from_owned_ptr(ptr))
        }
    }

    /// Metadata tag for the given codebook slot, if set.
    pub fn metadata_tag(&self, name: CodebookName) -> Option<MetadataTag> {
        // SAFETY: self is non-null and valid for the lifetime of `&self`.
        let ptr = unsafe {
            ffi::dnv_vista_sdk_local_id_builder_metadata_tag(
                self.as_ffi_ptr(),
                codebook_names::to_ffi(name),
            )
        };
        std::ptr::NonNull::new(ptr).map(MetadataTag)
    }

    /// Get the Quantity metadata tag (`None` if not set)
    pub fn quantity(&self) -> Option<MetadataTag> {
        self.metadata_tag(CodebookName::Quantity)
    }

    /// Get the Content metadata tag (`None` if not set)
    pub fn content(&self) -> Option<MetadataTag> {
        self.metadata_tag(CodebookName::Content)
    }

    /// Get the Calculation metadata tag (`None` if not set)
    pub fn calculation(&self) -> Option<MetadataTag> {
        self.metadata_tag(CodebookName::Calculation)
    }

    /// Get the State metadata tag (`None` if not set)
    pub fn state(&self) -> Option<MetadataTag> {
        self.metadata_tag(CodebookName::State)
    }

    /// Get the Command metadata tag (`None` if not set)
    pub fn command(&self) -> Option<MetadataTag> {
        self.metadata_tag(CodebookName::Command)
    }

    /// Get the Type metadata tag (`None` if not set)
    pub fn r#type(&self) -> Option<MetadataTag> {
        self.metadata_tag(CodebookName::Type)
    }

    /// Get the Position metadata tag (`None` if not set)
    pub fn position(&self) -> Option<MetadataTag> {
        self.metadata_tag(CodebookName::Position)
    }

    /// Get the Detail metadata tag (`None` if not set)
    pub fn detail(&self) -> Option<MetadataTag> {
        self.metadata_tag(CodebookName::Detail)
    }

    /// Returns a new builder with the VIS version set to `vis_version`.
    pub fn with_vis_version(&self, vis_version: VisVersion) -> Result<LocalIdBuilder, VistaError> {
        let c_version = CString::new(vis_version.as_str()).expect("invalid UTF-8 in VisVersion");
        // SAFETY: self and c_version are non-null/valid for the call's duration. Returns an owned pointer or NULL.
        let ptr = unsafe {
            ffi::dnv_vista_sdk_local_id_builder_with_vis_version(
                self.as_ffi_ptr(),
                c_version.as_ptr(),
            )
        };
        std::ptr::NonNull::new(ptr)
            .map(LocalIdBuilder)
            .ok_or_else(last_error)
    }

    /// Returns a new builder with the VIS version cleared.
    pub fn without_vis_version(&self) -> LocalIdBuilder {
        // SAFETY: self is non-null and valid for the lifetime of `&self`. Returns an owned pointer.
        let ptr =
            unsafe { ffi::dnv_vista_sdk_local_id_builder_without_vis_version(self.as_ffi_ptr()) };
        LocalIdBuilder(
            std::ptr::NonNull::new(ptr)
                .expect("dnv_vista_sdk_local_id_builder_without_vis_version returned NULL"),
        )
    }

    /// Returns a new builder with the primary item set to `path`.
    pub fn with_primary_item(&self, path: &crate::core::gmod_path::GmodPathRef) -> LocalIdBuilder {
        // SAFETY: self and path are non-null and valid for the call's duration. Returns an owned pointer.
        let ptr = unsafe {
            ffi::dnv_vista_sdk_local_id_builder_with_primary_item(
                self.as_ffi_ptr(),
                path.as_ffi_ptr(),
            )
        };
        LocalIdBuilder(
            std::ptr::NonNull::new(ptr)
                .expect("dnv_vista_sdk_local_id_builder_with_primary_item returned NULL"),
        )
    }

    /// Returns a new builder with the primary item cleared.
    pub fn without_primary_item(&self) -> LocalIdBuilder {
        // SAFETY: self is non-null and valid for the lifetime of `&self`. Returns an owned pointer.
        let ptr =
            unsafe { ffi::dnv_vista_sdk_local_id_builder_without_primary_item(self.as_ffi_ptr()) };
        LocalIdBuilder(
            std::ptr::NonNull::new(ptr)
                .expect("dnv_vista_sdk_local_id_builder_without_primary_item returned NULL"),
        )
    }

    /// Returns a new builder with the secondary item set to `path`.
    pub fn with_secondary_item(
        &self,
        path: &crate::core::gmod_path::GmodPathRef,
    ) -> LocalIdBuilder {
        // SAFETY: self and path are non-null and valid for the call's duration. Returns an owned pointer.
        let ptr = unsafe {
            ffi::dnv_vista_sdk_local_id_builder_with_secondary_item(
                self.as_ffi_ptr(),
                path.as_ffi_ptr(),
            )
        };
        LocalIdBuilder(
            std::ptr::NonNull::new(ptr)
                .expect("dnv_vista_sdk_local_id_builder_with_secondary_item returned NULL"),
        )
    }

    /// Returns a new builder with the secondary item cleared.
    pub fn without_secondary_item(&self) -> LocalIdBuilder {
        // SAFETY: self is non-null and valid for the lifetime of `&self`. Returns an owned pointer.
        let ptr = unsafe {
            ffi::dnv_vista_sdk_local_id_builder_without_secondary_item(self.as_ffi_ptr())
        };
        LocalIdBuilder(
            std::ptr::NonNull::new(ptr)
                .expect("dnv_vista_sdk_local_id_builder_without_secondary_item returned NULL"),
        )
    }

    /// Returns a new builder with `tag` set in the corresponding codebook slot.
    pub fn with_metadata_tag(
        &self,
        tag: &crate::core::metadata_tag::MetadataTagRef,
    ) -> LocalIdBuilder {
        // SAFETY: self and tag are non-null and valid for the call's duration. Returns an owned pointer.
        let ptr = unsafe {
            ffi::dnv_vista_sdk_local_id_builder_with_metadata_tag(
                self.as_ffi_ptr(),
                tag.as_ffi_ptr(),
            )
        };
        LocalIdBuilder(
            std::ptr::NonNull::new(ptr)
                .expect("dnv_vista_sdk_local_id_builder_with_metadata_tag returned NULL"),
        )
    }

    /// Returns a new builder with the metadata tag for `name` cleared.
    pub fn without_metadata_tag(&self, name: CodebookName) -> LocalIdBuilder {
        // SAFETY: self is non-null and valid for the lifetime of `&self`. Returns an owned pointer.
        let ptr = unsafe {
            ffi::dnv_vista_sdk_local_id_builder_without_metadata_tag(
                self.as_ffi_ptr(),
                codebook_names::to_ffi(name),
            )
        };
        LocalIdBuilder(
            std::ptr::NonNull::new(ptr)
                .expect("dnv_vista_sdk_local_id_builder_without_metadata_tag returned NULL"),
        )
    }

    /// Returns a new builder with verbose mode set to `verbose`.
    pub fn with_verbose_mode(&self, verbose: bool) -> LocalIdBuilder {
        // SAFETY: self is non-null and valid for the lifetime of `&self`. Returns an owned pointer.
        let ptr = unsafe {
            ffi::dnv_vista_sdk_local_id_builder_with_verbose_mode(
                self.as_ffi_ptr(),
                verbose as c_int,
            )
        };
        LocalIdBuilder(
            std::ptr::NonNull::new(ptr)
                .expect("dnv_vista_sdk_local_id_builder_with_verbose_mode returned NULL"),
        )
    }

    /// Constructs the final [`LocalIdRef`](crate::core::local_id::LocalIdRef). Returns `Err` if the builder state is not valid.
    pub fn build(&self) -> Result<LocalId, VistaError> {
        // SAFETY: self is non-null and valid for the lifetime of `&self`. Returns an owned pointer.or NULL.
        let ptr = unsafe { ffi::dnv_vista_sdk_local_id_builder_build(self.as_ffi_ptr()) };
        if ptr.is_null() {
            Err(last_error())
        } else {
            Ok(LocalId::from_owned_ptr(ptr))
        }
    }
}

impl PartialEq for LocalIdBuilderRef {
    fn eq(&self, other: &Self) -> bool {
        // SAFETY: self and other are non-null and valid for the lifetime of `&self`/`&other`.
        unsafe {
            ffi::dnv_vista_sdk_local_id_builder_equals(self.as_ffi_ptr(), other.as_ffi_ptr()) != 0
        }
    }
}

impl Eq for LocalIdBuilderRef {}

impl std::fmt::Display for LocalIdBuilderRef {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        // SAFETY: self is non-null and valid for the lifetime of `&self`. Returns an owned pointer.
        let ptr = unsafe { ffi::dnv_vista_sdk_local_id_builder_to_string(self.as_ffi_ptr()) };
        assert!(
            !ptr.is_null(),
            "dnv_vista_sdk_local_id_builder_to_string returned NULL"
        );
        // SAFETY: ptr is non-null and allocated by the library; invalid UTF-8 is replaced lossily.
        unsafe { crate::ffi::core::common::display_owned_cstr(ptr, f) }
    }
}

/// An owned, independently-released [`LocalIdBuilderRef`].
///
/// Returned by [`LocalIdBuilder::create`] and other APIs that hand back
/// a new builder value rather than a reference into existing state.
pub struct LocalIdBuilder(pub(crate) std::ptr::NonNull<ffi::dnv_vista_sdk_local_id_builder_t>);

impl Drop for LocalIdBuilder {
    fn drop(&mut self) {
        // SAFETY: self.0 is owned by this `LocalIdBuilder` and freed exactly once.
        unsafe { ffi::dnv_vista_sdk_local_id_builder_free(self.0.as_ptr()) }
    }
}

impl std::ops::Deref for LocalIdBuilder {
    type Target = LocalIdBuilderRef;

    fn deref(&self) -> &LocalIdBuilderRef {
        // SAFETY: self.0 is NonNull, so as_ptr() is always non-null and valid for `self`'s lifetime.
        unsafe { LocalIdBuilderRef::from_ptr(self.0.as_ptr()) }
    }
}

impl std::fmt::Display for LocalIdBuilder {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        std::fmt::Display::fmt(&**self, f)
    }
}

impl LocalIdBuilder {
    /// Create a new LocalIdBuilderRef for the given VIS version
    pub fn create(vis_version: VisVersion) -> Self {
        let c_version = CString::new(vis_version.as_str()).expect("invalid UTF-8 in VisVersion");
        // SAFETY: c_version is a valid NUL-terminated C string. Returns a non-null owned pointer.
        let ptr = unsafe { ffi::dnv_vista_sdk_local_id_builder_create(c_version.as_ptr()) };
        LocalIdBuilder(
            std::ptr::NonNull::new(ptr)
                .expect("dnv_vista_sdk_local_id_builder_create returned NULL"),
        )
    }

    /// Parses a LocalIdRef string. Returns `None` on invalid input.
    pub fn from_string(local_id_str: &str) -> Option<LocalId> {
        let c_str = CString::new(local_id_str).ok()?;
        // SAFETY: c_str is a valid NUL-terminated C string. Returns an owned pointer or NULL.
        let ptr = unsafe { ffi::dnv_vista_sdk_local_id_builder_from_string(c_str.as_ptr()) };
        if ptr.is_null() {
            None
        } else {
            Some(LocalId::from_owned_ptr(ptr))
        }
    }

    /// Parses a LocalIdRef string, returning the result together with any parse errors.
    pub fn from_string_with_errors(local_id_str: &str) -> (Option<LocalId>, ParsingErrors) {
        let c_str = CString::new(local_id_str).expect("local_id_str contains a NUL byte");
        let mut errors_ptr = ptr::null_mut();
        // SAFETY: c_str is a valid NUL-terminated C string, and `errors_ptr` is a valid
        // writable pointer to a local that the callee always sets.
        let ptr = unsafe {
            ffi::dnv_vista_sdk_local_id_builder_from_string_with_errors(
                c_str.as_ptr(),
                &mut errors_ptr,
            )
        };
        assert!(
            !errors_ptr.is_null(),
            "dnv_vista_sdk_local_id_builder_from_string_with_errors did not set outErrors"
        );
        let local_id = if ptr.is_null() {
            None
        } else {
            Some(LocalId::from_owned_ptr(ptr))
        };
        (
            local_id,
            ParsingErrors(std::ptr::NonNull::new(errors_ptr).expect("errors_ptr must not be NULL")),
        )
    }
}

// SAFETY: LocalIdBuilderRef is a borrowed view (`repr(transparent)`). Sync only, not Send.
unsafe impl Sync for LocalIdBuilderRef {}

// SAFETY: LocalIdBuilder owns its heap-allocated C++ object exclusively.
unsafe impl Send for LocalIdBuilder {}
unsafe impl Sync for LocalIdBuilder {}
